jtuples/src/main/java/org/jtuples/Tuple.java

```diff
@@ -0,0 +1,108 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.jtuples;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Represents a mathematical tuple.
+ * This is a fixed-length, immutable container of elements.
+ * The number and type of its members are determined by the implementing
+ * classes.
+ *
+ * @author Andre Santos <contact.andre.santos@gmail.com>
+ */
+public interface Tuple {
+    /**
+     * Returns the arity of this tuple.
+     * The arity of a tuple is the number of elements it contains.
+     * @return the arity of this tuple
+     */
+    int arity();
+
+    /**
+     * Returns a new tuple, with the same elements, in reverse order.
+     * @return a tuple in reverse order
+     */
+    Tuple invert();
+
+    /**
+     * Returns a {@code List} view of the elements in this tuple.
+     * The returned list is immutable.
+     * @return a list representation of this tuple
+     */
+    List<?> asList();
+
+    /**
+     * Returns an array containing the elements in this tuple.
+     * @return an array containing the elements in this tuple
+     */
+    Object[] toArray();
+
+    /**
+     * Compares the specified object with this tuple for equality.
+     * Returns {@code true} if the specified object is also a tuple,
+     * the two tuples have the same arity, and all corresponding pairs of
+     * elements in the two tuples are equal.
+     * (Two elements {@code e1} and {@code e2} are equal if
+     * {@code (e1 == null ? e2 == null : e1.equals(e2))}.)
+     * In other words, two tuples are defined to be equal if they contain
+     * the same elements in the same order.
+     * This definition ensures that the {@code equals} method works properly
+     * across different implementations of the {@code Tuple} interface.
+     * @param obj object to be compared for equality with this tuple
+     * @return true if the specified object is equal to this tuple
+     */
+    @Override
+    boolean equals(Object obj);
+
+    /**
+     * Returns the hash code value for this tuple.
+     * The hash code of a tuple is defined to be the result of the following
+     * calculation.
+     * <pre>{@code
+     *     int hashCode = 1;
+     *     for (Object e: tuple.asList())
+     *         hashCode = 31*hashCode + (e==null ? 0 : e.hashCode());}
```

Comment:
It somewhat standard for hash codes. I just followed the recipe from `List`, since that formula takes order into account. It isn't _necessary_, though. Should it be ammended, just to state that it should consider all elements and their order?